fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java

```diff
@@ -255,4 +402,57 @@ public interface PlayChannelHandler {
 		 */
 		void receive(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender responseSender);
 	}
+
+	/**
+	 * An internal packet handler that works as a proxy between old and new API.
+	 * @param <T> the type of the packet
+	 */
+	private interface PlayChannelHandlerProxy<T extends FabricPacket> extends PlayChannelHandler {
+		PlayPacketHandler<T> getOriginalHandler();
+	}
+
+	/**
+	 * A thread-safe packet handler utilizing {@link FabricPacket}.
+	 * @param <T> the type of the packet
+	 */
+	@FunctionalInterface
+	public interface PlayPacketHandler<T extends FabricPacket> {
+		/**
+		 * Handles the incoming packet. This is called on the render thread, and can safely
+		 * call rendering-related methods.
+		 *
+		 * <p>An example usage of this is to display an overlay message:
+		 * <pre>{@code
+		 * // See FabricPacket for creating the packet
+		 * ClientPlayNetworking.registerReceiver(OVERLAY_PACKET_TYPE, (player, packet, responseSender) -> {
+		 * 	MinecraftClient.getInstance().inGameHud.setOverlayMessage(packet.message(), true);
+		 * });
+		 * }</pre>
+		 *
+		 * <p>The network handler can be accessed via {@link ClientPlayerEntity#networkHandler}.
+		 *
+		 * @param client the client
+		 * @param player the player that received the packet
+		 * @param packet the packet
+		 * @param responseSender the packet sender
+		 * @see FabricPacket
+		 */
+		void receive(MinecraftClient client, ClientPlayerEntity player, T packet, PacketSender responseSender);
+
+		/**
+		 * Processes an incoming packet. This is called on the network thread.
+		 * If this method returns {@code false}, {@link #receive} is not called.
+		 * This is for advanced users only; for normal uses, {@link #receive} should be
+		 * implemented instead.
+		 *
+		 * @param client the client
+		 * @param player the player that received the packet
+		 * @param packet the packet
+		 * @param responseSender the packet sender
+		 * @return whether to proceed to {@link #receive}
```

Comment:
This is unclear and needs to specify explicitly what happens with `{@code true}` / `{@code false}`. Even then, the name could be misleading (one could understand it as "should processing happen on network thread"). A less ambiguous name would be nice.